### ISSUE TITLE
imxrt-multi/uart, rt1064 - fd, RX is not working on other UARTs than the console.

### ISSUE THREAD

**Issue by gerard5** [phoenix-rtos/phoenix-rtos-devices#55]
Two UARTs are used in my case (1st: default phoenix-rtos/multi-driver console and 3rd: non-blocking, rw, HDLC pipe [termios=as much raw as possible, no echo]).
The console TX/RX works, uart-3 only TX works (RX worked with a small amount of simultaneous raw data up to four data words being received without the idle spacing, more than 4 simultaneous data words caused the hw FIFO to enter the over-run state and required at least (OR) status flag to be cleared and some times (FE) flag too to re-enable the operational state of the receiver. After the kernel update from #f465316 to #8676efa the receive is not working on other UART's than the console which in my case is uart-1. What is good, after the kernel update console started to accept more than four simultaneous characters (e.g. by copy&paste of many commands to the terminal) without single FIFO overrun and terminal hiccups, which is good sign - possibly due to changes in the kernels armv7/interrupts-imxrt. Additionally setting libtty/termios with tcsetattr() produces: Usage Fault exception after the kernel update.
As opposed to file descriptors, using message communication (mtRead/Write) uart-3 works RX/TX but O_NONBLOCK has no effect and blocks both either TX or RX but reception of data is possible, and I almost forgot (mtDevCtl) or an ioctl to libtty produces Usage Fault as with tcsetattr() mentioned before, so I'm unable to change termios/libtty settings and baud rate.

**Comment by nalajcie** [comment]
1. **Use some formatting** for better visibility (You can always use the "preview" tab to see if it look ok before submitting).

2. **Please provide minimal steps to reproduce the issue** (eg. no need for HDLC to debug UART). Maybe a short code for setting up the uart and reading the received bytes would be helpful - if you have it - to quickly reproduce the issue. Otherwise provide **steps** as a numbered list of actions to be done (eg. provide command You use to setup uart3). Right now we're missing some information regarding the UART speed, HW control (RTS/CTS pin usage) usage, etc.

3. Tip: for referencing commits from different repos either hand-write the referenced commit, eg: phoenix-rtos/phoenix-rtos-kernel@f465316 or paste the full github URL (pasting `https://github.com/phoenix-rtos/phoenix-rtos-kernel/commit/8676efa8d57db48358c801a28aa6f2159caa7cd2` will result in: https://github.com/phoenix-rtos/phoenix-rtos-kernel/commit/8676efa8d57db48358c801a28aa6f2159caa7cd2). If You're referencing commit from the current repo, just a short SHA is needed - so writing `1b0e8ee` would result in 1b0e8ee.

**Comment by gerard5** [comment]
Thank You for the quick response, to be correct as much as possible I've checked everything again and rephrased the problem.

To reproduce the issues described above please use the following code (I focused on the file descriptor approach rather than messages because of the portability):
https://gist.github.com/gerard5/a56ac3fcef5eda4bfd3420e6250f7662

Target is the **NXP MIMIXRT1064-EVK** evaluation platform, ; so please connect (**uart-3**, which is available on the ,,arduino'' socket **J22** [pin 1 is the RX, pin 2 is the TX], check if jumpers R288 & R289 are soldered or propagated with 0R) with the PC using USB-UART transceiver. The RTS/CTS is not used, only RX/TX. The system console shall be available on (**uart-1**) with build in NXP UART to USB transceiver, to build the kernel I've used the following setup:
```
KERNEL_CFLAGS=" -DUART_CONSOLE=1"
DEVICES_CFLAGS=" -DUART_CONSOLE=1 -DUART1=1 -DUART3=1 -DUART3_TX_PIN=ad_b1_06 -DUART3_RX_PIN=ad_b1_07"
```
and after the kernel update, multi-map setup is as bellow
```
MULTI_MAPS="M0x20018000;0x20028000;0 M0x20220000;0x2027fe00;0"
SYSPAGE_PROGS="imxrt-multi;1" "dummyfs;2" "uart3echo_tst;2" "psh;2"
```
The program (published as gist above) is echoing the 512 octet buffer, to feed it please use virtually any tool on the PC side, e.g. **socat** with **sha256sum** to verify the correctness of echoed data. On the PC, I'm using OpenBSD actually, but with Linux it should be easy available too.

I work with the latest 8d46223 changes as I'm writing this. An update of the kernel I meant jump from phoenix-rtos/phoenix-rtos-kernel@f465316 to phoenix-rtos/phoenix-rtos-kernel@8676efa so I have not tried any of the commits in the middle. The latest available kernel as for now is not stable, so I keep with what's used in other phoenix projects which are current (call stable kernel), available through grit as super repo with submodules and being around the rt1064 target.

1. Worked here phoenix-rtos/phoenix-rtos-kernel@f465316, not working there phoenix-rtos/phoenix-rtos-kernel@8676efa
If baud rate is set to other speed than 115.2k the `tcsetattr(fd, TCSANOW, &tio)` in line 72 crashes:
```
exception: 6 #UsageFault
 r0=00000000  r1=00000000  r2=00000000  r3=00000000
 r4=00000002  r5=0000000f  r6=20019a28  r7=00000001
 r8=20018190  r9=20018000 r10=20019a28 r11=0000000f
r12=00000000  sp=20008f2c  lr=00000000  pc=00000000
psp=20018cb0 psr=00000000 exr=ffffffff bfa=00000000
cfs=00020000
```
The current project expect to work with 230kbps, and maximum PDU of 510 bytes, with phoenix-rtos/phoenix-rtos-kernel@f465316 I was able to set 460800 and do TX from rt1064 to PC at that baud rate, but no RX as described above.

2. Doesn't work with phoenix-rtos/phoenix-rtos-kernel@f465316, but partly works with phoenix-rtos/phoenix-rtos-kernel@8676efa.
- Doesn't work mean the situation when hw RX consumed 4 bytes, fireing IRQs (I've used GPIO drived from IRQ and a bunch of LEDs to see the state of status flags) but threads conditional was not unlocked before hardware fifo entered the over-run state (OR), and eventually set (FE) status flag, needed clearing appropriate flags in the IRQ handler (OR, FE) to re-enable receiver.

- Partly works means that after change in the kernel, see the diff between the two hashes: (reschedule and priorities of SYSTICK, PENDSV, SVC_IRQ), hardware fifo doesn't over-run anymore, but libtty produces a lot `libtty-disc: RX OVERRUN!` messages which are in software and related to imx-multi/uart/libtty, I've tried to change the buffer size with no effect. Please see that data received with (read) in the program above is chunked to portions of ~20 octets (which is expected to be around that value), so it should not over-run the libtty as its buffer is at least 256 bytes, or previously 512.

**Comment by agkaminski** [comment]
I've managed to reproduce your problem. I did it on master of all project's repositories. Found stack overflow in libtty, most likely fixed in phoenix-rtos/phoenix-rtos-devices@40462e7b574f5257500a71177adedaa082fa2767, please confirm on your side.

**Comment by pawelpisarczyk** [comment]
Gerard, is this problem solved by Alek? Could you comment on that?

**Comment by gerard5** [comment]
The solution provided by Alek does not solve the problem. I talked, about the issue described here, with Alek, we agreed that I should try with the latest kernel, which I was able to build, program the rt1064 with psu and run, but I still get Usage Faults - only when trying to change baud rate (ioctl/**tcsetattr**). Additionaly, when debug is enabled I get `libtty-disc: RX OVERRUN!` on receiving bulk characters (circa 100), while libtty BUFSIZE is set to 512 characters.
